pkg/fuzzer/fuzzer.go

```diff
@@ -68,67 +65,105 @@ func NewFuzzer(ctx context.Context, cfg *Config, rnd *rand.Rand,
 	return f
 }
 
-type Config struct {
-	Debug          bool
-	Corpus         *corpus.Corpus
-	Logf           func(level int, msg string, args ...interface{})
-	Coverage       bool
-	FaultInjection bool
-	Comparisons    bool
-	Collide        bool
-	EnabledCalls   map[*prog.Syscall]bool
-	NoMutateCalls  map[int]bool
-	FetchRawCover  bool
-	NewInputFilter func(call string) bool
+type execQueues struct {
+	smashQueue           *queue.PlainQueue
+	triageQueue          *queue.PriorityQueue
+	candidateQueue       *queue.PlainQueue
+	triageCandidateQueue *queue.PriorityQueue
+	source               queue.Source
 }
 
-type Request struct {
-	Prog       *prog.Prog
-	NeedSignal SignalType
-	NeedCover  bool
-	NeedHints  bool
-	// If specified, the resulting signal for call SignalFilterCall
-	// will include subset of it even if it's not new.
-	SignalFilter     signal.Signal
-	SignalFilterCall int
-	// Fields that are only relevant within pkg/fuzzer.
-	flags   ProgTypes
-	stat    *stats.Val
-	resultC chan *Result
+func newExecQueues(fuzzer *Fuzzer) execQueues {
+	ret := execQueues{
+		triageCandidateQueue: queue.Priority(),
+		candidateQueue:       queue.PlainWithStat(fuzzer.StatCandidates),
+		triageQueue:          queue.Priority(),
+		smashQueue:           queue.Plain(),
+	}
+	// Sources are listed in the order, in which they will be polled.
+	ret.source = queue.Multiplex(
+		ret.triageCandidateQueue,
```

Comment:
Looking how processResult is done synchronously in one case, but asynchronously using OnDone in another case, and how prepare is called manually in genFuzz (so can be forgotten to be called in some other place).

With the flexible queues architecture, perhaps we can do something along the following lines to ensure consistent handling of each request:
```
ret.source = queue.Wrapper(ret.source, fuzzer.wrapRequest)
....
func (fuzzer) wrapRequest(req *Request) {
  varify req
  install fuzzer's OnDone callback
}
```

Reply:
The problem is that `fuzzer.Fuzzer` installs somewhat different OnDone callbacks - with different sets of `execOpt`.

Reply:
I see. Pity we can't easily associate arbitrary data with Request. If ProgType would be part of Request, then it won't be a problem, we could have a the same generic callback.
But not sure how to do it nicely. Putting `map[any]any` into Request for anybody to add anything looks like overkill, inefficient and may be subject to races.
We could add a single Flags field into Request, ProgType and noTriage fits into it. Manager will also need to associate "stared executing" flag with requests.

Reply:
>  Putting map[any]any into Request for anybody to add anything looks like overkill, inefficient and may be subject to races.

Theoretically, we could make use of Go's https://pkg.go.dev/context#WithValue by putting a context.Context to each `Request`. Not sure it's worth it though.

Reply:
Context looks even less convenient.